routing/http/types/ndjson/provider.go

```diff
@@ -0,0 +1,96 @@
+package ndjson
+
+import (
+	"encoding/json"
+	"io"
+
+	"github.com/ipfs/go-libipfs/routing/http/types"
+	"github.com/ipfs/go-libipfs/routing/http/types/iter"
+)
+
+type readProvidersResponseIter struct {
+	iter.Iter[types.UnknownProviderRecord]
+}
+
+func NewReadProvidersResponseIter(r io.Reader) *readProvidersResponseIter {
+	return &readProvidersResponseIter{Iter: iter.FromReaderJSON[types.UnknownProviderRecord](r)}
+}
+
+func (p *readProvidersResponseIter) Next() (types.ProviderResponse, bool, error) {
+	v, ok, err := p.Iter.Next()
+	if err != nil {
+		return nil, false, err
+	}
+	if !ok {
+		return nil, false, nil
+	}
+	switch v.Schema {
+	case types.SchemaBitswap:
+		var prov types.ReadBitswapProviderRecord
+		err := json.Unmarshal(v.Bytes, &prov)
+		if err != nil {
+			return nil, false, err
+		}
+		return &prov, true, nil
```

Comment:
Another example how having only two params simplifies iterator implementations and it is less error prone (there is no chance to flip an ok boolean when is not the case):

```go
return json.Unmarshal(...)
```